### ISSUE TITLE
closure variable modifications are losing in unittest

### ISSUE THREAD

**Issue by ymcvalu** [golang/go#53697]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.17.11 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
i can't reproduce on go1.18.3

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env

GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOOS="darwin"
GOPRIVATE=""
GOTMPDIR=""
GOVCS=""
GOVERSION="go1.17.11"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
</pre></details>

### What did you do?

```go
func Recover(ctx context.Context, cbs ...func(ctx context.Context, err error)) {
	r := recover()
	if r == nil {
		return
	}

	err, is := r.(error)
	if !is {
		err = fmt.Errorf("panic with err: %v", r)
	} else {
		err = fmt.Errorf("panic with err: %w", err)
	}

	log.Stack(ctx, "goroutine panic", zap.Error(err))
	for _, cb := range cbs {
		cb(ctx, err)
	}
}

func LP() (err error) {
	ctx := context.Background()
	cb := func(ctx context.Context, pe error) {
		err = pe
		log.Info(ctx, "===> cb run", zap.Error(err))
	}
	defer Recover(ctx, cb)

	panic("test")
}
```

```go 
func main() {
	fmt.Println("===>", LP())
}
```
---
```go
func TestLP(t *testing.T) {
	err := LP()
	t.Log("===>", err, err == nil)
}
```

### What did you expect to see?

the `LP` func should returns a non-nil error

### What did you see instead?

when i using `go run main.go`, the `LP` func returns a non-nil error,
but i get a nil error when using `go test -v -count=1 -run=TestLP`

**Comment by ZekeLu** [comment]
This issue has been fixed as a side effect of 0a0e3a3dea72d8d64d4250c9f7649da3b942eae5, which has been shipped in go1.18.

**Comment by heschi** [comment]
Go 1.17 will be out of support when 1.19 ships in about a month, so it's unlikely we'd backport a fix to it now. I suggest you update to 1.18 or 1.19 when it comes out.
